\begin{abstract}
Since its emergence almost 20 years ago (Schwartz et al., Science 1995), optical mapping has undergone a transition from laboratory technique to commercially available data generation method. In line with this transition, 
it is only relatively recently that optical mapping data has started to be used for scaffolding contigs and assembly validation in large-scale sequencing projects --- for example, the goat (Dong et al., Nature Biotech. 2013) and amborella (Chamala et al., Science 2013) genomes. 
One major hurdle to the wider use of optical mapping data
is the efficient alignment of {\em in silico} digested contigs to an optical map. We develop $\twin$ to tackle this very problem.  $\twin$ is the first index-based method for aligning {\em in silico} digested contigs to an optical map.  Our results demonstrate that $\twin$ is an order of magnitude faster than competing methods on the largest genome. Most importantly, it is specifically designed to be capable of dealing with large eukaryote genomes and thus is the only non-proprietary method capable of completing the alignment for the budgerigar genome in a reasonable amount of CPU time.   
\end{abstract}
