
\begin{table*}[t]
\centering

\begin{tabular}{| 
			p {0.22\linewidth} |
			p {0.20\linewidth} |
			p {0.15\linewidth} |
			p {0.13\linewidth} |
			p {0.25\linewidth} | }
			
			%l|c|c|c|}
\hline
{\bf Genome} 			&  {\bf Program}	& {\bf Memory }	& {\bf Time } 			& {\bf Aligned Contigs} \\ 

\hline
\hline
{\em Y. Kristensenii} & & &  & \\
\hline

				& Valouev {\em et al.} 	& 1.81 		& .17 s 			& 91  \\
				& SOMA 				& 1.71 		& 7.32 s 			& 54 \\
				& $\twin$
                                & 18  		& .06 s
                                & 65\\
% 61 acheived with search radius 1000 and cvalue of 4
\hline
\hline
Rice & & &   & \\
\hline 

				& Valouev {\em et al.} 	& 11.25 		& 2 m 57 s 			& 2,676  \\
				& SOMA 				& 7.94 		& 29 m 38 s 		& 2,434 \\
				& $\twin$
                                & 18.25  		&  50 s 			&  3,098\\
% 2439 acheived with search radius 1600 and C value of 10
\hline
\hline
Budgerigar & & & & \\
\hline 

				& Valouev {\em et al.} 	& 390  			& 6.5 h 		& 9,814 \\
				& SOMA 				& 380.95  		& 77.2 d 		& 9,668 \\
				& $\twin$                         &127.112                  &  35 m           & 9,826\\
% 8915 achieved with bin size 150, search radius of 1500, and C value
% of 200

\hline
\end{tabular}
\caption{{\bf Comparsion of the alignment results for $\twin$ and competing method.}  The performance of $\twin$ was compared against SOMA \cite{Nagarajan08} and the method of Valouev et al.~\cite{Valouev06} using the assembly and optical mapping data for {\em Yersinia Kristensenii}, rice, and budgerigar.  Various assemblers were used to assemble the data for these species.  The relevant statistics and information concerning these assemblies and genomes can be found in Table \ref{tab:assembly_stats}.  The peak memory is given in megabytes (mb).  The running time is reported in seconds (s), minutes (m), hours (h), and days. }
\label{tab:possible_columns}
\end{table*} 

