
%%%%%%%%%%%%%%%%%%%%%%% file typeinst.tex %%%%%%%%%%%%%%%%%%%%%%%%%
%
% This is the LaTeX source for the instructions to authors using
% the LaTeX document class 'llncs.cls' for contributions to
% the Lecture Notes in Computer Sciences series.
% http://www.springer.com/lncs       Springer Heidelberg 2006/05/04
%
% It may be used as a template for your own input - copy it
% to a new file with a new name and use it as the basis
% for your article.
%
% NB: the document class 'llncs' has its own and detailed documentation, see
% ftp://ftp.springer.de/data/pubftp/pub/tex/latex/llncs/latex2e/llncsdoc.pdf
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%\documentclass[runningheads,a4paper]{llncs}
\documentclass[11pt,letterpaper]{article}


\usepackage{epsfig,multicol}
\usepackage{graphicx}
\usepackage{amsthm}
\usepackage{amssymb}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{mathrsfs}


\usepackage{url}
\usepackage{algorithm}
\usepackage{amssymb,amsmath}
\usepackage{algorithmicx}
\usepackage[noend]{algpseudocode}
\usepackage{sidecap}
\setcounter{tocdepth}{3}
\usepackage{graphicx}
\usepackage{color}

\def\ST{\mbox{\rm {\sf ST}}}
%\def\ST{\mbox{\rm {\sf M}}}
\def\SA{\mbox{\rm {\sf SA}}}
\def\ISA{\mbox{\rm {\sf ISA}}}
\def\LZ{\mbox{\rm {\sf LZ}}}
\def\LF{\mbox{\rm {\sf LF}}}
\def\FMI{\mbox{\rm {\sf FMI}}}
\def\FM{\mbox{\rm {\sf FM}}}
%\def\rank{\mbox{\rm {\sf rank}}}
\def\rank{\textsf{rank}}
\def\select{\textsf{select}}
\def\ralpha{\textsf{ralpha}}
\def\pathlabel{\mbox{\rm {\sf pathlabel}}}
\def\lcp{\mbox{\rm {\sf lcp}}}
\def\rmq{\mbox{\rm {\sf rmq}}}
\def\x{\mbox{\rm {\sf x}}}
%\def\X{\mbox{\rm {\sf X}}}
\def\X{\mathsf{X}}
\def\MS{\mbox{\rm {\sf MS}}}
%\def\B{\mbox{\rm {\sf B}}}
\def\B{\mathsf{B}}
\def\ssB{\mbox{\rm {\sf {\scriptsize B}}}}
\def\A{\mathsf{A}}
%\def\A{\mbox{\rm {\sf A}}}
\def\Pr{\mbox{\rm {\sf P}}}
\def\R{\mbox{\rm {\sf R}}}
\def\C{\mbox{\rm {\sf C}}}

\def\Y{\mbox{\rm {\sf Y}}}
\def\Y{\mathsf{Y}}
\def\ssY{\mbox{\rm {\sf {\scriptsize Y}}}}
%\def\Z{\mbox{\rm {\sf Z}}}
\def\Z{\mathsf{Z}}
\def\C{\mbox{\rm {\sf C}}}
\def\D{\mbox{\rm {\sf D}}}
\def\M{\mbox{\rm {\sf M}}}
\def\F{\mbox{\rm {\sf F}}}
\def\BWT{\mbox{\rm {\sf BWT}}}

\def\twin{\mbox{\rm {\sc Twin}}}
\def\dopp{\mbox{\rm {\sc doppelganger}}}

\usepackage{url}


\newcommand{\keywords}[1]{\par\addvspace\baselineskip
\noindent\keywordname\enspace\ignorespaces#1}


%\usepackage{vmargin}
%\setmarginsrb{1in}{1in}{1in}{1in}{0pt}{0pt}{0pt}{6mm}
\begin{document}



\title{Fast and accurate Rmap alignment}
\author{Martin D. Muggli\thanks{Department of Computer Science,Colorado State University, Fort Collins, CO}
\and Simon J. Puglisi\thanks{Department of Computer Science,University of Helsinki, Finland}
\and Christina Boucher\footnotemark[1]}%~\thanks{Corresponding author.}
\date{}
\maketitle

\begin{abstract}
%A dearth of computational methods for handling optical mapping data currently hampers its wider use. 
An optical map provides genome-wide positional information
for a set of short nucleotide sequences. Because they are derived
independently of short read data, optical maps can aid de novo assemblers
and/or be used to effectively cross check and correct assemblies.
However, a dearth of computational methods for handling optical mapping
data currently hampers its wider use.
Indeed, as Mendelowitz and 
Pop (GigaScience 2014) recently emphasise:  ``[t]here is, thus, a critical need for the continued 
development and public release of software tools for processing optical mapping data, mirroring the 
tremendous advances made in analytical methods for second- and third-generation sequencing data.''  
One fundamental computational task in the use of optical mapping data is the assembly of the genome-wide 
optical map from single molecule optical map sequences, called Rmaps, which are to a genome-wide optical map what short 
reads are to a completed genome sequence. 
%that can be used for a variety of purposes.  
%The first step of the assembly process --- computing all pairwise alignments between the Rmaps --- 
%is a significant algorithmic challenge in itself. 
We present an index-based, error-tolerant 
alignment method that can be used for pairwise alignment (and a variety of other purposes) and demonstrate it is currently the only public method capable of performing pairwise 
alignment for Rmaps derived from the plum genome in a reasonable amount of time.
\end{abstract}

%\begin{abstract}
%Mendelowitz and Pop (GigaScience 2014) state that:  ``There is, thus, a critical need for the continued development and public release of software tools for processing optical mapping data, mirroring the tremendous advances made in analytical methods for second- and third-generation sequencing data..''  One of the major impediments of the use of optical mapping data is the assembly of the raw data (called Rmaps) into a genome wide optical map that can be used for a variety of purposes.  The first step of the assembly process is computing all pairwise alignments between the Rmaps and is a significant algorithmic challenge in it of itself.   We present an index-based, error-tolerant alignment method that can be used for this said purpose and a variety of other purposes.  Our experimental results demonstrate that our method is currently the only method that can complete the pairwise alignment for the goat Rmaps in a reasonable amount of time.
%\end{abstract}

\input{./content/introduction}
\input{./content/background}
\input{./content/methods}
\input{./content/results}
\input{./content/discussion}\

%\section*{Acknowledgements} 
%The authors would like to thank Journi Sir\'{e}n from the Wellcome Trust Sanger Institute for many insightful discussions concerning the GCSA code base.   MM and CB were funded by the Colorado Clinical and Translational Sciences Institute which is funded by National Institutes of Health (NIH-NCATS,UL1TR001082, TL1TR001081, KL2TR001080).  SJP was supported by the Helsinki Institute of Information Technology (HIIT) and by Academy of Finland through grants 258308 and 250345 (CoECGR).  All authors greatly appreciate the funding provided for this project.  


\bibliographystyle{splncs_srt}
\bibliography{./content/gcsa_paper,sjp}

\end{document}
