\section{Conclusion}
\label{p3-sec:conclusion}

We presented $\ours$, which is an implementation of a succinct colored de Bruijn graph that significantly reduces the amount of memory required to store and use the colored de Bruijn graph. In addition to the memory savings, we validated our approach using {\em E coli.} Moreover, we introduced the use of colored de Bruijn graph for accurately identifying the presence of AMR genes within metagenomic samples, which is an important advance as public health officials increasingly move towards a metagenomic sequence-based approach for surveillance and identification of resistant bacteria \cite{baquero_metagenomic_epi, port_2014_metagenomics_AMR_monitoring,FAOActionPlan2016}. Possible nontrivial extensions to our work include (1)
% fully scalable is no longer future work since we do (and describe) this now.  -MDM
% developing a fully scalable version of our construction algorithm that makes use of external-memory and parallelized sorting;
(1) using multi-threading to speed up the bubble calling, (2) compressing the color array $C$ more effectively by taking advantage of correlations among the variations, and (3) applying more sophisticated approaches to metagenomic data.


%Nontrivial extensions to our work include (1) developing a fully scalable version of our construction algorithm that makes use of external-memory sorting, and (2) determining a succinct data structure that would allow for efficient querying of large metagenomics sample datasets.
%A tailored colored de Bruijn graph implementation that would enable efficiently identify and comparison of AMR genes and their sequence variations across thousands of samples would be an influential method for AMR research.


%% we can't put this in, it would be embarrassing:  A trivial but important change to our implementation is to unbound the number of colors in order to handle large populations

%Another efficient implementation exists called Bloom-filter trie~\cite{BFT}, but is heavily engineered for applications different from those considered in the Cortex paper.Due to time constraints, we have not investigated whether and how it can be optimized for our experiments. A fair comparison with $\ours$ presents another focus for future work.
