

\documentclass[doctor]{thesis}



\usepackage[cmex10]{amsmath}
\usepackage{amsthm,amssymb}
\usepackage{flafter}
\usepackage[pdftex]{graphicx} \graphicspath{{./figures/}} 
\DeclareGraphicsExtensions{.pdf,.png,.jpg}

%\usepackage[caption=false]{subfig}

\usepackage{booktabs}

\usepackage{url}
\urlstyle{same} 

% \title{Resource constrained genomics via succinct data structures: Complex variants exposed as features in massive sample sets for sample origin classification}

\title{Enhancing Space and Time Efficiency of Genomics in Practice Through Sophisticated Applications of the FM-Index}

\author{Martin D. Muggli}

\email{martin.muggli@colostate.edu}

\department{Department of Computer Science}

\semester{Fall 2018}

\advisor{Ross McConnell}
%\coadvisor{Co-Advisor Name}
\committee{Charles Anderson}
\committee{Hamid Chitsaz}
\committee{Paul S Morley}

% This document contains text and art co-authored with others and whose copyright has been assigned to Springer or Oxford University Press.  It is reproduced with permission. All other content, 
\mycopyright{Copyright by Martin D. Muggli 2018 \\
All Rights Reserved
}
% both all other co-authors and  from  copyright holders in compliance with university requirements allowing non-original work.  

\abstract{Genomic sequence data has become so easy to get that the computation to process it has become a bottleneck in the advancement of biological science.  A data structure known as the FM-Index both compresses data and allows efficient querying, thus can be used to implement more efficient processing methods.  In this work we apply advanced formulations of the FM-Index to existing problems and show our methods exceed the performance of competing tools.   }

% \abstract{  Modern genome sequencing is largely based on a process of randomly breaking replicated copies of a genome into fragments, using various technologies to capture the nucleotide sequence within these fragments (resulting in strings known as reads), and then using assembly software to attempt to reconstruct the original genome sequence from the reads.
% This process is challenging as genomes contain repeated regions, and repeated regions much longer than read length confound assemblers, limiting their ability to completely and correctly reconstruct genomes successfully.
% Correct and complete genome assembly is important because genomes encode elements that cooperate with others in close proximity, and thus not just the content, but  genome structure has important biological implications.
% To the extent quality automated genome reconstruction is possible, there is an additional challenge of accessibility, as some of the most successful assembly software requires unusually high-end servers or clusters.
% This limits their usefulness to biologists with access and skill to use such machines and hence more efficient computational techniques are of value.
% Beyond efficiency and correctness of algorithms, there is interplay between computational approach, sequencing technology (which vary in read length, accuracy, applicability, and level of detail), and the assembly quality that may result.
% In this report, we will expand on the concepts introduced here and review a selection of modern computational assembly tools, the sequence data on which they operate, and discuss important advantages, limitations, and possible extensions of them as well as their relationship to each other in the context of the sequence assembly problem.
% }


\acknowledgements{I would like to thank Christina Boucher who filled the roll of advisor for all of this work for her patience, guidance, and insight in my training. I would also like to thank my committee, Ross McConnell, Charles Anderson, Hamid Chitsaz, and Paul S Morley, for their continued help and support. I would like to thank the many people who have supported me emotionally throughout this work, specifically: Alex Bowe, Lindsay Burchfield, Aaron Ciuffo, Katie Davidson, Teri Dillion, Thomas Harrison, Amber Johnson, Anne Mangiardi, Piper Murray, Simon Puglisi, Susan Rainey, Sanjay Rajopadhye, Sara Rector, Stacey Reynolds, Basir Shariat, Mike Storlie, Michelle Strout, John Wagner, Jason Walp, Adrienne Watral and others who I may have forgotten.  I would also like to thank the various people and organizations  whose outstanding science work provided inspiration: Richard P Feynman, ESA, CERN, LIGO, GHC. In addition, I would like to thank my co-authors for their contribution to our work. And last but not least I would like to thank my family, Jaynie Muggli, Jim Muggli, and Karen Muggli for their support as well.
}


\usepackage[pdfpagelabels,pdfusetitle,colorlinks=false,pdfborder={0 0 0}]{hyperref}

\input{defs}
\begin{document} 
\frontmatter 
\maketitle
\makemycopyright
\makeabstract
\makeacknowledgements   
% \prelimtocentry{Dedication} \begin{flatcenter} 
%         DEDICATION

%         \vfill 
% %    \noindent \textit{I would like to dedicate this thesis to my dog fluffy.}
%     \vfill \end{flatcenter}
\newpage

\tableofcontents
%\listoftables       \listoffigures      
\mainmatter 
